Add an async socket.io API

1c3t3a opened the pull request:
This PR builds upon the async engine.io version and adds an async socket.io interface.
The interface is similar to the sync one and provides most methods in an async fashion. Note that the async version can currently only be used with `tokio`.

An Issue I stumbled upon while writing this, is the verbose way of defining callbacks (they need to return boxed futures).
I explained the issue [here](https://users.rust-lang.org/t/type-inference-issue-with-closures-returning-a-future/74088). From my point of view there is currently no better way than using `async { .. }.boxed()`.

nshaaban-cPacket commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -0,0 +1,733 @@
+        Ok(())
+    }
+
+    // TODO: 0.3.X add secure socketio server
```
nit: Is this test a blocker on 0.4.0 coming out of alpha?

1c3t3a replied:
I don't think so

nshaaban-cPacket commented on socketio/src/asynchronous/generator.rs:
```diff
@@ -0,0 +1,34 @@
+use futures_util::{ready, FutureExt, Stream, StreamExt};
+use tokio::sync::Mutex;
+
+pub(crate) type Generator<T> = Pin<Box<dyn Stream<Item = T> + 'static + Send>>;
```
nit: a comment explaining why this is needed would be nice. Very much appreciate it being it's own type.

nshaaban-cPacket commented on socketio/src/asynchronous/socket.rs:
```diff
@@ -0,0 +1,228 @@
+
+#[derive(Clone)]
+pub(crate) struct Socket {
+    //TODO: 0.4.0 refactor this
```
nit: why does it need refactoring?

nshaaban-cPacket commented on socketio/src/asynchronous/socket.rs:
```diff
@@ -0,0 +1,228 @@
+    pub(super) fn new(engine_client: EngineClient) -> Result<Self> {
+        let connected = Arc::new(AtomicBool::default());
+        Ok(Socket {
+            engine_client: Arc::new(engine_client.clone()),
+            connected: connected.clone(),
+            generator: StreamGenerator::new(Self::stream(engine_client, connected)),
```
nit: Cloning this doesn't seem safe, might result in undefined behavior. We could end up with two independent clients. Might cause issues for secure socket io websocket for instance

nshaaban-cPacket replied:
Cloning an object only guarantees a copy of the present state, but depending on the implementation of clone/and the struct the behavior could either be:
- Two rust objects pointing to the same data using arc/locks
- Two rust objects that are copies of the same data when they're cloned, but changes to one isn't reflected into the other (if we don't have guarantees that engine_client does this it could result in strange behavior)

1c3t3a replied:
That's true, for the engine_client we know that this works as all necessary parts live behind an `Arc`. But I see the smell, this is not obvious. I will think about another way.

1c3t3a replied:
Good idea, for now I added a test and updated documentation. But I think I'll change that in the future.

nshaaban-cPacket commented on socketio/src/asynchronous/socket.rs:
```diff
@@ -0,0 +1,228 @@
+        self.send(socket_packet).await
+    }
+
+    /// Returns a packet for a payload, could be used for bot binary and non binary
```
typo: bot -> both

nshaaban-cPacket reviewed (changes requested):
main concerns:
- consider async callback in a separate feature
- tests don't test to make sure the callbacks function
- version bump

1c3t3a commented on socketio/src/asynchronous/client/builder.rs:
```diff
@@ -0,0 +1,329 @@
+    pub fn namespace<T: Into<String>>(mut self, namespace: T) -> Self {
+        let mut nsp = namespace.into();
+        if !nsp.starts_with('/') {
+            nsp = "/".to_owned() + &nsp;
```
I don't feel well about logging, but I'm going to write a doc comment indicating the behavior. That'll show up on the documentation for the method.

1c3t3a commented on socketio/src/asynchronous/client/builder.rs:
```diff
@@ -0,0 +1,329 @@
+                for packet in socket_clone.next().await {
+                    if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = packet {
+                        //TODO: 0.3.X handle errors
+                        panic!("{}", e.unwrap_err());
```
This panic the entire application actually... I know it's annoying but I don't really know what to do here. If the underlying transport is in an irreparable state, all future messages / actions are invalid. As this runs in the background I can't really report the issue to the user (or only if this method returns a channel, which is also a weird interface...)

nshaaban-cPacket replied:
blocker: panicking libraries is a nightmare, especially one such as this where I (as the caller) can't prevent that situation from occurring.

IMO: log (other libraries such as actix log at the trace level) + throw an error packet

1c3t3a replied:
Oke, I added the tracing. The error also shows up in case the user polls the stream. I don't feel good about it as this might lead to weird situations, but you are right about the panic.

1c3t3a commented on socketio/src/lib.rs:
```diff
@@ -97,16 +176,18 @@ pub(self) mod socket;
 
 pub use client::{Client, ClientBuilder, TransportType};
 
-// TODO: 0.4.0 remove
-#[deprecated(since = "0.3.0-alpha-2", note = "Socket renamed to Client")]
-pub use client::{Client as Socket, ClientBuilder as SocketBuilder};
```
Yeah I thought so, but I don't want to crash your plans for the async server (if I recall correctly the server release is 0.4.0).

nshaaban-cPacket replied:
Burn 'em! I'm not married to those numbers, update the version as it makes sense for this PR and the library as a whole.

1c3t3a commented on socketio/src/lib.rs:
```diff
@@ -74,6 +74,85 @@
+//! ## Async version
+//! This library provides an ability for being executed in an asynchronous context using `tokio` as
+//! the execution runtime.
+//! Please note that the current async implementation is in beta, the interface can be object to
+//! drastic changes.
```
Yes that's the plan!

1c3t3a commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -0,0 +1,733 @@
+    }
+
+    /// Connects the client to a server. Afterwards the `emit_*` methods can be
+    /// called to interact with the server. Attention: it's not allowed to add a
+    /// callback after a call to this method.
```
Ah this is actually from an old version of the client. Now the builder ensures I only add a callback at the beginning.

nshaaban-cPacket replied:
nit: remove outdated comment

1c3t3a commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -0,0 +1,733 @@
+    async fn handle_event(&self, packet: &Packet) -> Result<()> {
+        // unwrap the potential data
+        if let Some(data) = &packet.data {
+            // the string must be a valid json array with the event at index 0 and the
+            // payload at index 1. if no event is specified, the message callback is used
```
That's necessary because of the way `serde_json` parses a String.

nshaaban-cPacket replied:
If possible prefer JSON objects over arrays
```
{
event: 0,
payload: ""
}
```

1c3t3a replied:
Ah sorry, just realized this is in fact a socketio requirement. A standard message for a certain event is encoded as the following JSON object: `["test_event", "test_msg"]`.

1c3t3a commented on socketio/Cargo.toml:
```diff
@@ -23,6 +23,22 @@ byte = "0.2.4"
 thiserror = "1.0"
```
Sounds reasonable

1c3t3a commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -0,0 +1,733 @@
+    Event, Payload,
+};
+
+/// Represents an `Ack` as given back to the caller. Holds the internal `id` as
+/// well as the current ack'ed state. Holds data which will be accessible as
+/// soon as the ack'ed state is set to true. An `Ack` that didn't get ack'ed
+/// won't contain data.
+#[derive(Debug)]
+pub struct Ack {
+    pub id: i32,
+    timeout: Duration,
+    time_started: Instant,
+    callback: Callback,
+}
```
This is only an internal type (passed around to make sure we've got all information for an `Ack`). I'll move it to another file.

1c3t3a commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -0,0 +1,733 @@
+            // poll for the next payload
+            let next = ready!(self.socket.poll_next_unpin(cx));
+
+            if let Some(result) = next {
```
Oh fair point!

1c3t3a commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -0,0 +1,733 @@
+    }
+
+    async fn test_socketio_socket(mut socket: Client, nsp: String) -> Result<()> {
+        let _: Option<Packet> = Some(socket.next().await.unwrap()?);
```
This is actually the opening packet.

1c3t3a commented:
Current open points:
- fix tests so that it is ensured that callbacks are called correctly
- investigate UB that might arise when cloning the client
- put async callbacks behind a feature flag

1c3t3a commented:
@nshaaban-cPacket What alternatives to async callbacks would you consider? A possible alternative would be to somehow use the implementation of `Stream` to react to a packet and provide means to answer to the server directly. I don't like the way the callbacks work for a two main reasons:
- The fact that the callback takes a `Client` as an argument is the root of the whole "client-needs-to-be-clonable" problem
- `Fn`s are hard to handle (especially in async context, where creation of `Fn`s is difficult and counterintuitive) also they need to be boxed

1c3t3a commented:
I added the `async-callbacks` feature flag. As there is currently no other alternative to async callbacks, I added it as a default feature when `async` is enabled. Anyhow the `ClientBuilder`'s `on<F: Fn(...) -> ...>(..)` method is excluded when the flag is not enabled, giving an option to add different versions of the method (or other interfaces like trait implementations) as needed.

nshaaban-cPacket commented:
> @nshaaban-cPacket What alternatives to async callbacks would you consider? A possible alternative would be to somehow use the implementation of `Stream` to react to a packet and provide means to answer to the server directly.

Either:
- streams (good in theory, bit of a pain to implement)
- trait based callback. not exactly sure how this works internally, but the idea is provide a trait that implements handle_message (or similar), then pass that object into the client. then users can create their own structs with their own state that handle socket.io messages.

nshaaban-cPacket commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -300,38 +289,46 @@ impl Client {
+                &event,
+                contents
+                    .get(1) // case 1: payload lives at index 1
+                    .unwrap_or_else(|| contents.get(0).unwrap()) // case 2: payload lives at index 0
```
suggestion: safety? may want to `let payload = contents.get(1);`

nshaaban-cPacket commented:
Sorry it took so long to review, didn't see that you re-requested review.

> fix tests so that it is ensured that callbacks are called correctly

May want to remove those tests that aren't actually testing so the coverage number is accurate.

nshaaban-cPacket commented on socketio/src/asynchronous/client/client.rs:
```diff
@@ -303,30 +303,29 @@ impl Client {
+                (
+                    contents
+                        .get(0)
+                        .map(|value| match value {
+                            Value::String(ev) => ev,
+                            _ => "message",
+                        })
+                        .unwrap_or("message")
+                        .into(),
```
```suggestion
                        .map(|value| match value {
                            Value::String(ev) => Event::from(ev),
                            _ => Event::Message,
                        })
                        .unwrap_or(Event::Message),
```

JackFlukinger commented:
Hey, FWIW, I tried out the async version and everything seemed to work except for event callbacks (not sure comprehensively these are tested). The sync version of the library works fine with the same calling code, so I'm pretty sure it was not a server config problem.

SSebo commented:
@JackFlukinger I found the callback not await, so not really executed. @1c3t3a  I made a PR #226

SSebo commented:
> @nshaaban-cPacket What alternatives to async callbacks would you consider? A possible alternative would be to somehow use the implementation of `Stream` to react to a packet and provide means to answer to the server directly. I don't like the way the callbacks work for a two main reasons:
> 
> * The fact that the callback takes a `Client` as an argument is the root of the whole "client-needs-to-be-clonable" problem

https://github.com/SSebo/rust-socketio/blob/0b1890e5f7036105ef6661c4be7656af43f0ef16/socketio/src/client/client.rs#L37-L54

hi, @1c3t3a . For reconnecting, the Client is cloneable, and the real client is `inner: Arc<RwLock<Client>>`. because we need to create a new socket to replace failed one, so the whole "client-needs-to-be-clonable" problem is acceptable I think.

nshaaban-cPacket commented:
> What alternatives to async callbacks would you consider?

Another option is rather than sending the client in the callback, send an intermediary object that can uses a message channel (or similar means) to send messages out over the client. Inside the callback you don't really need to have the full powers of the client, just send messages/close connection/etc.

1c3t3a commented:
Thanks for the valuable discussion in this PR! I am going to merge this PR now, it comes with fairly good documentation of the new async part and I hope to get some feedback on the async API eventually (especially regarding the valuable points here raised by @SSebo and @nshaaban-cPacket). I am releasing this as an `alpha` version, since we want to be flexible with API changes, the async part is behind a feature flag either way :) Thanks to everyone that was involved in this, whether it was through testing or reviewing this PR! :)
